### ISSUE TITLE
Allow overload parameter name mismatches if the implementation allows all possible call signatures

### ISSUE THREAD

**Issue by Numerlor** [microsoft/pyright#2630]
**Is your feature request related to a problem? Please describe.**
When the implementation of overloads uses parameter names different to the overloads, an error is raised for mismatched names, even if the signatures are compatible through other parameters.

For example this code fails because the name of the implementation's first parameter does not match the name of the parameter in the two overloads
```py
import typing as t

_MISSING_SENTINEL: t.Any = object()

@t.overload
def f(a: int, b: str):
    ...

@t.overload
def f(b: str):
    ...

def f(a_or_b: t.Union[str, int] = _MISSING_SENTINEL, /, b: str = _MISSING_SENTINEL, *, a: int = _MISSING_SENTINEL):
    ...
```
This signature does conform to the interfaces described by the overloads, but is not accepted for either of the overloads because `a_or_b` doesn't match `a` and `b`.

If the names are fixed (along with adjusting the signature), no errors are raised for the code, however this makes the API poorer as the names from the overloads won't necessarily make as much sense if they have to match. 
An another solution would be to make all of the parameters position or keyword only, but this again restricts the usability a bit.

**Describe the solution you'd like**
The implementation of overloads is accepted if the names mismatch, as long as the signature provides a keyword only parameter with a default value to accept the value if it is supplied as a kwarg.

**Comment by erictraut** [comment]
Pyright is correct to emit an error in this case. Not surprisingly, mypy also considers this an error. The overload signatures that you've provided are not compatible with the implementation.

There are legal combinations of overloads for this implementation signature. Consider the following:

```python
@t.overload
def f(a_or_b: t.Union[str, int] = _MISSING_SENTINEL, /):
    ...

@t.overload
def f(a_or_b: str = _MISSING_SENTINEL, *, a: int):
    ...

@t.overload
def f(a_or_b: int = _MISSING_SENTINEL, *, b: str):
    ...

@t.overload
def f(*, a: int, b: str):
    ...
```

If you don't want to do that, you can use a `# type: ignore` to suppress the error.

**Comment by Numerlor** [comment]
> Pyright is correct to emit an error in this case. Not surprisingly, mypy also considers this an error. The overload signatures that you've provided are not compatible with the implementation.

In what way are they incompatible? As I understand it, the implementation does allow all of the possible ways to be called that are described be the overload. Then for the others it allows it's up to the code to (not) handle

**Comment by Numerlor** [comment]
I've realised `*args, **kwargs` would pass this, so it seems strange to allow arbitrary arguments, but not a signature that is a restricted superset of the overloads
